core/src/main/scala/zio/jdbc/JdbcEncoder.scala

```diff
@@ -16,623 +16,271 @@
 
-  implicit val intEncoder: JdbcEncoder[Int]                               = value => sql"$value"
-  implicit val longEncoder: JdbcEncoder[Long]                             = value => sql"$value"
-  implicit val doubleEncoder: JdbcEncoder[Double]                         = value => sql"$value"
-  implicit val charEncoder: JdbcEncoder[Char]                             = value => sql"$value"
-  implicit val stringEncoder: JdbcEncoder[String]                         = value => sql"$value"
-  implicit val booleanEncoder: JdbcEncoder[Boolean]                       = value => sql"$value"
-  implicit val bigIntEncoder: JdbcEncoder[java.math.BigInteger]           = value => sql"$value"
-  implicit val bigDecimalEncoder: JdbcEncoder[java.math.BigDecimal]       = value => sql"$value"
-  implicit val bigDecimalEncoderScala: JdbcEncoder[scala.math.BigDecimal] = value => sql"$value"
-  implicit val shortEncoder: JdbcEncoder[Short]                           = value => sql"$value"
-  implicit val floatEncoder: JdbcEncoder[Float]                           = value => sql"$value"
-  implicit val byteEncoder: JdbcEncoder[Byte]                             = value => sql"$value"
-  implicit val byteArrayEncoder: JdbcEncoder[Array[Byte]]                 = value => sql"$value"
-  implicit val byteChunkEncoder: JdbcEncoder[Chunk[Byte]]                 = value => sql"$value"
-  implicit val blobEncoder: JdbcEncoder[java.sql.Blob]                    = value => sql"$value"
-  implicit val uuidEncoder: JdbcEncoder[java.util.UUID]                   = value => sql"$value"
-
-  implicit def singleParamEncoder[A: SqlFragment.Setter]: JdbcEncoder[A] = value => sql"$value"
+  def apply[A](
+    onEncode: A => SqlFragment,
+    onSql: => String,
+    onValue: (PreparedStatement, Int, A) => Unit,
+    onNull: (PreparedStatement, Int) => Unit
+  ): JdbcEncoder[A] =
+    new JdbcEncoder[A] {
 
-  // TODO: review for cases like Option of a tuple
-  def optionEncoder[A](implicit encoder: JdbcEncoder[A]): JdbcEncoder[Option[A]] =
-    value => value.fold(SqlFragment.nullLiteral)(encoder.encode)
+      override def encode(value: A): SqlFragment = onEncode(value)
 
-  implicit def tuple2Encoder[A: JdbcEncoder, B: JdbcEncoder]: JdbcEncoder[(A, B)] =
-    tuple => JdbcEncoder[A]().encode(tuple._1) ++ SqlFragment.comma ++ JdbcEncoder[B]().encode(tuple._2)
+      override def unsafeSetValue(ps: PreparedStatement, index: Int, value: A): Int = {
+        onValue(ps, index, value)
+        index + 1
+      }
 
-  implicit def tuple3Encoder[A: JdbcEncoder, B: JdbcEncoder, C: JdbcEncoder]: JdbcEncoder[(A, B, C)] =
-    tuple =>
-      JdbcEncoder[A]().encode(tuple._1) ++ SqlFragment.comma ++ JdbcEncoder[B]().encode(
-        tuple._2
-      ) ++ SqlFragment.comma ++ JdbcEncoder[C]().encode(tuple._3)
+      override def unsafeSetNull(ps: PreparedStatement, index: Int): Int = {
+        onNull(ps, index)
+        index + 1
```

Comment:
Again, `onNull` must return an int, because what if we are talking about a composite JdbEncoder, e.g. Person(name, age), and then when you set that to null, you actually have to set 2 placeholders to null.